troposphere/static/js/actions/instance/resume.js

```diff
@@ -1,58 +1,90 @@
+ */
+function handleOne(instance) {
+    // if we've been passed a model via "selected resources" that
+    // is *not* an Instance, then just skip it - _move on_
+    if (!(instance instanceof Instance)) return;
+
+    let instanceState = new InstanceState({
+            status_raw: "suspended - resuming",
+            status: "active",
+            activity: "resuming"
+        }),
+        originalState = instance.get("state"),
+        actionRequest = new InstanceActionRequest({
+            instance: instance
+        });
+
+    instance.set({
+        state: instanceState
+    });
+
+    Utils.dispatch(InstanceConstants.UPDATE_INSTANCE, {
+        instance: instance
+    });
+
+    actionRequest.save(null, {
+        attrs: {
+            action: "resume"
+        }
+    }).done(function() {
+        // Consider "success" notification ...
         instance.set({
             state: instanceState
         });
+    }).fail(function(response) {
+        instance.set({
+            state: originalState
+        });
+        Utils.displayError({
+            title: "Your instance could not be resumed",
+            response: response
+        });
+    }).always(function() {
         Utils.dispatch(InstanceConstants.UPDATE_INSTANCE, {
             instance: instance
         });
-
-        actionRequest.save(null, {
-            attrs: {
-                action: "resume"
-            }
-        }).done(function() {
-            instance.set({
-                state: instanceState
-            });
-        }).fail(function(response) {
-            instance.set({
-                state: originalState
-            });
-            Utils.displayError({
-                title: "Your instance could not be resumed",
-                response: response
-            });
-        }).always(function() {
-            Utils.dispatch(InstanceConstants.UPDATE_INSTANCE, {
-                instance: instance
-            });
-            Utils.dispatch(InstanceConstants.POLL_INSTANCE_WITH_DELAY, {
-                instance: instance,
-                delay: 25*1000,
-            });
+        Utils.dispatch(InstanceConstants.POLL_INSTANCE_WITH_DELAY, {
+            instance: instance,
+            delay: 25*1000,
         });
+    });
+}
+
+
+export default {
+
+    resume: function(params) {
+        if (!params.resources)
+            throw new Error("Missing resources");
+        if (params.instance)
+            throw new Error("Deprecated parameter passed");
+
+        // We expect `resources` to be a backbone collection
+        let resources = params.resources;
+
+        // in case "resources" is not a collection, but is
+        // an Instance passed as `resources`, make a collection
+        // and add the "models/Instance" to it
+        if (resources instanceof Instance) {
+            let instance = resources;
+            resources = new Backbone.Collection(instance);
+        }
+
+        resources.map(handleOne);
```

Comment:
I would just like to mention that all I did here was an _Extract Method_ refactor to existing, _redundant_ code. 

I think that the single files `{stop,start,resume,suspend}` could be rolled up into a single module with a common function that exposed the action-specifics as parameters. 

Something like this:

```
// we define a "function" constructor that will create a closure for an actionName
// and the closure will take one argument, the instance
function instanceAction(actionName, desiredState, errorMsg, delayOptions) {
    delayOptions = delayOptions || { delay: 25*1000 };

    return function(instance) {
        let instanceState = new InstanceState(desiredState),
        originalState = instance.get("state"),
        actionRequest = new InstanceActionRequest({
            instance: instance
        });

        instance.set({
            state: instanceState
        });

        Utils.dispatch(InstanceConstants.UPDATE_INSTANCE, {
            instance: instance
        });

        actionRequest.save(null, {
            attrs: {
                action: actionName
            }
        }).done(function() {
            instance.set({
                state: instanceState
            });
        }).fail(function(response) {
            instance.set({
                state: originalState
            });
            Utils.displayError({
                title: errorMsg,
                response: response
            });
        }).always(function() {
            Utils.dispatch(InstanceConstants.UPDATE_INSTANCE, {
                instance: instance
            });
            Utils.dispatch(InstanceConstants.POLL_INSTANCE_WITH_DELAY, {
                instance: instance,
                delay: delayOptions.delay,
            });
        });
    }
}
```

And then `resume`, `suspend`, `stop`, `start`, (even `shelve`) become named functions that we can use within the modals:

```
// imagine we have a module named "errorMessages" ... 

const instanceResume =  instanceAction(
    "resume", 
    {
            status_raw: "suspended - resuming",
            status: "active",
            activity: "resuming"
    }, 
    errorMessages.resumeFailed
);

const instanceStop =  instanceAction(
    "stop", 
    {
            status_raw: "active - powering-off",
            status: "active",
            activity: "powering-off"
    }, 
    errorMessages.stopFailed
);

const instanceStart =  instanceAction(
    "start", 
    {
            status_raw: "shutoff - powering-on",
            status: "shutoff",
            activity: "powering-on"
    }, 
    errorMessages.startFailed
);

const instanceSuspend =  instanceAction(
    "suspend", 
    {
            status_raw: "active - suspending",
            status: "active",
            activity: "suspending"
    }, 
    errorMessages.suspendFailed
);

export { instanceResume, instanceSuspend, instanceStop, instanceStart }
```

The reason for `errorMessages` is that we are going to move toward i18n & l10n soon, so we could start by extracting the error message text into a module now, and then the messages can be handled at a later time. 

@cdosborn - is this at all in line with how you were thinking about avoiding repetition?